src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,118 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size;
+    private Object[] data;
+
+    public ArrayList() {
+        this(DEFAULT_CAPACITY);
+    }
+
+    public ArrayList(int capacity) {
+        if (capacity < 0) {
+            throw new IllegalArgumentException("Illegal Capacity: " + capacity);
+        }
+        data = new Object[capacity];
+    }
+
     @Override
     public void add(T value) {
-
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't add element. Illegal index: "
+                    + index);
+        }
+        if (size == data.length) {
+            grow(size + 1);
+        }
+        if (size > index) {
+            System.arraycopy(data, index, data, index + 1,size - index);
+        }
+        data[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (data.length < size + list.size()) {
+            grow(size + list.size());
+        }
+        for (int i = 0; i < list.size(); i++) {
+            this.add(list.get(i));
+        }
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't get element. Illegal index: "
+                    + index);
+        }
+        return (T) data[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't set element. Illegal index: "
+                    + index);
+        }
+        data[index] = value;
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't remove element. Illegal index: "
+                    + index);
+        }
+        T element = (T) data[index];
+        if (--size >= index) {
+            System.arraycopy(data, index + 1, data, index, size - index);
+        }
```

Comment:
```suggestion
     
            System.arraycopy(data, index + 1, data, index, size - index);
        
```

Reply:
Removing this check, definitely wouldn't be great idea.

For some removeElementFromFullArrayListByIndex(core.basesyntax.ArrayListTest): arraycopy: last source index 11 out of bounds for object array[10]